### ISSUE TITLE
Status not saved when moving to rebirth zone

### ISSUE THREAD

**Issue by kyc2976** [aura-project/aura#378]
[youtube](https://youtu.be/M4pySlIQe9Y)

please watch that video.

Character window - Rebrith button - (status save) - rebirth zone (Soul Stream)

but **sometimes** status doesn't save.   item, skill rank and quest…. maybe sql data.

I'm sorry to my bad english.

**Comment by exectails** [comment]
We had such an issue before, but we weren't able to reproduce it and there was no indicator as to what might've gone wrong =/ I'll have to take a closer look at this, thanks for the report and the video.

**Comment by DennouNeko** [comment]
There's something resetting (or skipping setting) the Save flag when leaving Soul Stream...
I think the only method that sets this is the EnterRegionRequest handler and that's why it worked after moving around.
I had similar problem with secondary logins when I left the Account property being overwritten by new login.

**Comment by Pegwald** [comment]
I wonder if this might be related to a closed issue were I ranked an ability, rebirthed, then found the skill unranked after the rebirth. I can't remember if I lost the AP or not...

**Comment by DennouNeko** [comment]
Hm... For now just looked around a bit. So far:
- SaveAccount() is being called only from Client.CleanUp()
- Client.CleanUp() is being called directly only in DisconnectRequest handler or in Client.Kill() (which is called pretty much only in case of error), unless I've missed some indirect call.
- On rebirth client seems to send only SwitchChannel, but doesn't call DisconnectRequest.

**Comment by exectails** [comment]
Can anyone reproduce this? I can't seem to get it to happen.

@DennouNeko `CleanUp` is called from `KillAndRemoveClient` on disconnect, it should be called *always*, no matter how the connection is terminated, unless the server gets killed. Though I didn't know we're basically calling CleanUp twice... it has obviously worked for a while now, and shouldn't cause this issue, but it's probably not a good idea.

**Comment by DennouNeko** [comment]
Didn't actually check it, but from what I've seen in logs, I'm not sure if the connection is terminated. Client seems to keep the connection alive and just sends switch channel + new login. It's something to be verified.

**Comment by exectails** [comment]
It is terminated, that's the first thing I checked^^

**Comment by DennouNeko** [comment]
Ah... Then I'm out of ideas :p

**Comment by exectails** [comment]
I have a theory. If the client relogged faster than the server can save the progress on disconnect, the new connection would load the old state, while the actual state is still being saved. Since there's no reconnect on going from Soul Stream to the world, you keep the state you had when "you saved" last, since next time you log out, your actual state, that was saved when you were already in, gets overwritten.

This could happen sporadically, whenever the client manages to reconnect and log in before the server notices the disconnect of the client.

Basically, this would cause a rollback of your entire session before you clicked rebirth. Quite a problematic bug =< A hint to the fix is this:
>On rebirth client seems to send only SwitchChannel, but doesn't call DisconnectRequest.

We simply have to save on the switch channel request, which takes over 500ms in my logs, so officials probably do the same.

**Comment by exectails** [comment]
I've added the proposed fix, since I really can't think of any other reason why this might happen.

@kyc2976 It would be good if you could try to reproduce this bug again, so we can make sure it's fixed.
